mobilecoind-dev-faucet/src/slam/prepared_utxo.rs

```diff
@@ -0,0 +1,214 @@
+}
+
+impl PreparedUtxo {
+    /// Make a new prepared utxo
+    ///
+    /// Start with a UtxoRecord, and get proofs and mixins from mobilecoind
+    pub async fn new(
+        index: usize,
+        utxo_record: UtxoRecord,
+        params: &SlamParams,
+        mobilecoind_api_client: &MobilecoindApiClient,
+        logger: &Logger,
+    ) -> Result<PreparedUtxo, String> {
+        let mut tries = 1;
+        let (proofs_resp, mixins_resp) = loop {
+            match Self::get_proofs_and_mixins(&utxo_record, mobilecoind_api_client).await {
+                Ok(result) => {
+                    break result;
+                }
+                Err(err) => {
+                    log::debug!(
+                        logger,
+                        "Preparing tx #{} (attempt {}/{}) : {}",
+                        index,
+                        tries,
+                        params.retries,
+                        err
+                    );
+                    if tries == params.retries {
+                        log::error!(
+                            logger,
+                            "Failed preparing tx #{} on {} tries: {}",
+                            index,
+                            params.retries,
+                            err
+                        );
+                        return Err(err);
+                    }
+                    tokio::time::sleep(params.retry_period).await;
+                }
+            }
+            tries += 1;
+        };
+
+        // The Tx builder sorts the ring anyways, so it doesn't matter if we always put
+        // the real input first.
+        let (ring, membership_proofs): (Vec<TxOut>, Vec<TxOutMembershipProof>) = proofs_resp
+            .get_output_list()
+            .iter()
+            .chain(mixins_resp.get_mixins().iter())
+            .map(|tx_out_with_proof| {
+                Ok((
+                    tx_out_with_proof.get_output().try_into()?,
+                    tx_out_with_proof.get_proof().try_into()?,
+                ))
+            })
+            .collect::<Result<Vec<_>, ConversionError>>()
+            .map_err(|err| format!("Conversion error: {}", err))?
+            .into_iter()
+            .unzip();
+
+        Ok(PreparedUtxo {
+            index,
+            utxo_record,
+            ring,
+            membership_proofs,
+            real_input_index: 0,
+        })
+    }
+
+    async fn get_proofs_and_mixins(
+        utxo_record: &UtxoRecord,
+        mobilecoind_api_client: &MobilecoindApiClient,
+    ) -> Result<
+        (
+            mc_mobilecoind_api::GetMembershipProofsResponse,
+            mc_mobilecoind_api::GetMixinsResponse,
+        ),
+        String,
+    > {
+        // Get a membership proof for this utxo
+        let mut req = mc_mobilecoind_api::GetMembershipProofsRequest::new();
+        req.set_outputs(RepeatedField::from(vec![utxo_record
+            .utxo
+            .get_tx_out()
+            .clone()]));
```

Comment:
Optional/Equivalent: Use `push`
```suggestion
        req.mut_outputs().push(utxo_record.utxo.get_tx_out().clone());
```